src/storage/log/aof/log.rs

```diff
@@ -205,31 +231,44 @@ impl AOL {
         buf: &mut [u8],
         segment_id: u64,
         read_offset: u64,
-    ) -> io::Result<usize> {
-        if segment_id == self.active_segment_id {
+    ) -> Result<usize> {
+        // During read, we acquire a lock to not allow concurrent writes and reads
+        // to the active segment file to avoid seek errors.
+        if segment_id == self.active_segment.id {
+            let _lock = self.mutex.write();
             self.active_segment.read_at(buf, read_offset)
         } else {
-            let segment = Segment::open(&self.dir, segment_id, &self.opts)?;
-            segment.read_at(buf, read_offset)
+            let mut cache = self.segment_cache.write();
+            let reader = cache.get(&segment_id);
+            if reader.is_none() {
+                let segment = Segment::open(&self.dir, segment_id, &self.opts)?;
+                let read_bytes = segment.read_at(buf, read_offset)?;
+                cache.push(segment_id, segment);
+                Ok(read_bytes)
+            } else {
+                let segment = reader.unwrap();
+                segment.read_at(buf, read_offset)
+            }
         }
     }
 
-    pub fn close(&mut self) -> io::Result<()> {
-        let _lock = self.mutex.write().unwrap();
+    pub fn close(&mut self) -> Result<()> {
+        let _lock = self.mutex.write();
         self.active_segment.close()?;
         Ok(())
     }
 
-    pub fn sync(&mut self) -> io::Result<()> {
-        let _lock = self.mutex.write().unwrap();
-        self.active_segment.sync()?;
-        Ok(())
+    // Returns the current offset within the segment.
+    pub fn offset(&self) -> Result<u64> {
+        let _lock = self.mutex.read();
```

Comment:
Do we need to hold a read lock while reading `active_segment`? If so, shouldn't `active_segment` itself be protected by the mutex? Someone could just access it directly otherwise.

Reply:
So I've been using a mutex inside the struct to manage this, the expectation is for the end user to not use these internal fields at all, but rather the methods. The read lock just ensures the writer waits because these offsets are being used to track the position of the record in the file and don't want this to be a dirty read as such.

Reply:
> the expectation is for the end user to not use these internal fields at all, but rather the methods.

I wasn't worried about end users. My worry was that we could access the offsets from another part of the codebase directly. Could be simply a case of forgetting to acquire the mutex first or ignorance. Over time more people are probably going to contribute to this library, not just you, which increases the risk of this happening. Rust mutexes are designed to avoid this from happening by holding the data they are protecting so that the only way you can get to it is by acquiring the mutex first.

Reply:
I see your point, I'll submit a separate PR for this as I wasn't really sure on what pattern to use for accessing locks across multiple fields in Rust. Appreciate it